drud/ddev
# Custom host and in-container commands, fixes #1372, fixes #1551

## The Problem/Issue/Bug:

One of many things requested in #1372 and not yet finished was custom commands, so you could write a command with bash or however, and execute it with `dev <customcommandname>`

## How this PR Solves The Problem:

* Allow custom commands on the host, created with a simple script formula and placed in .ddev/commands/host
* Allow custom commands that do their work int he container, simple script in .ddev/commands/web


## Manual Testing Instructions:

* Read the docs. See [current docs](https://github.com/drud/ddev/blob/5fa05b2bb125455538e069e59ed7ee955e4a853c/docs/users/extend/custom-commands.md)
* Review the examples in your project in .ddev/commands/*/*
* Use the examples in .ddev/commands to experiment with both host-side and container-side commands
* Use `ddev` and `ddev <command> -h` to explore the custom commands 
* Use `ddev mysql`

## Automated Testing Overview:

* Added TestCmdCustomCommands, covers quite a bit.

## Related Issue Link(s):

#1372 is ongoing issue that covers this
#1551 requests the "ddev mysql" command, which is provided here.

## Release/Deployment notes:
